Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,640 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Starboard.h>
+#import <StubReturn.h>
+#import <Foundation/NSDecimal.h>
+#import <algorithm>
+
+#define IS_NEGATIVE(decimal) (decimal)->_isNegative
+
+#define IS_POSITIVE(decimal) !IS_NEGATIVE(decimal)
+
+#define IS_ZERO(decimal) (((decimal)->_length == 0) && IS_POSITIVE(decimal))
+
+// clang-format off
+#define INVALIDATE_DECIMAL(decimal) \
+decimal->_length = 0;               \
+decimal->_isNegative = YES;
+
+#define SET_TO_ZERO(decimal) \
+decimal->_length = 0;        \
+decimal->_isNegative = NO;   \
+NSDecimalCompact(decimal);
+// clang-format on
+
+#define VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand)                 \
+    if (NSDecimalIsNotANumber(leftOperand) || NSDecimalIsNotANumber(rightOperand)) { \
+        INVALIDATE_DECIMAL(result);                                                  \
+        return NSCalculationOverflow;                                                \
+    }
```

Comment:
Lets not use this pattern. Including a return statement in a macro is a dangerous game as it may not be clear to a casual reader what the full implications are and control path assumptions may prove to be invalid. 

In general I am not a fan of how many macros you are using here. From CC:

2.2.15 Use macros with discretion

Use macros only when they are absolutely necessary. Most functionality that is typically achieved through macros can be implemented through other C++ constructs (using constants, enums, inline functions, or templates) which will yield clearer, more resilient, more debuggable code.

[https://github.com/Microsoft/WinObjC/wiki/Coding-Standards](https://github.com/Microsoft/WinObjC/wiki/Coding-Standards)

Much of the MACRO use here simply obfuscates the true intent and doesn't save all that much code.